### PR TITLE
docs: Add design document for Control Plane Ingress HA

### LINKED ISSUE #2381: Control plane Ingress controller is tied to Bootstrap
**Component**: salt, ingress

**Problem**:

> The way the control-plane Ingress controller is currently deployed ain't great (using an `externalIP`).
> (see [this comment](https://github.com/scality/metalk8s/issues/2380#issuecomment-610396113))

This means the controller can only run on the Node with this `externalIP` value, in our case, the Bootstrap. (Note that the certificates are also only generated for this IP in their SANs)

**Resolution proposal**:

> This should be fixed somehow anyway (basically, your control-plane ingress running on a non-bootstrap node won't even work today). Either by having a specific `portmap` configuration, or by running them in the host network namespace and bind to the (host) control-plane IP only, somehow.
> (see [this comment](https://github.com/scality/metalk8s/issues/2380#issuecomment-610396113))

### PR DESCRIPTION
**Pull request by TeddyAndrieux**
Design for Control Plane Ingress HA

---

Sees: #2381

### CODE REVIEW

File: docs/developer/architecture/control-plane-ingress.rst [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+I can still access and authenticate on the MetalK8s and Grafana UIs.
+
+Design Choices
+--------------
+
+To have a proper HA Control Plane Ingress we want to use a
+`Virtual IP <https://en.wikipedia.org/wiki/Virtual_IP_address>`_ using
+`MetalLB <https://metallb.universe.tf/>`_ so that we can rely on layer2
+ARP requests when possible or let MetalLB advertise IPs using BGP when
+the user router support it.
```
Comment (NicolasT) [inline docs/developer/architecture/control-plane-ingress.rst]: Do we plan to support BGP-based address management?

File: docs/developer/architecture/control-plane-ingress.rst [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+NOTE: MetalLB will be used only for Control Plane Ingress and not
+for Workload Plane.
```
Comment (NicolasT) [inline docs/developer/architecture/control-plane-ingress.rst]: Never ever?

File: docs/developer/architecture/control-plane-ingress.rst (hunk 1 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: I don't know, it's more a question for @thomasdanan 
But to me, since it's built-in metallb it's only some field to fill in the metallb config it's easy to "support" it

File: docs/developer/architecture/control-plane-ingress.rst (hunk 2 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: Hmm at least it's not part of this specific design, I can change a bit this sentence

File: docs/developer/architecture/control-plane-ingress.rst [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+- Using a user-provided IP that should switch between Control
+  Plane nodes (managed by the user)
+
+NOTE: In all those 3 approaches we want the user to provide the Control
+Plane Ingress IP he wants to use.
+
+Rejected Design Choices
+-----------------------
+
+TODO
```
Comment (NicolasT) [inline docs/developer/architecture/control-plane-ingress.rst]: Please do before merging :wink:

File: docs/developer/architecture/control-plane-ingress.rst [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+re-configured.
+
+NOTE: Changing this Control Plane Ingress IP means we need to reconfigure
+all Kubernetes APIServer since we use this Ingress IP as an OIDC provider.
+
+MetalLB Configuration
+~~~~~~~~~~~~~~~~~~~~~
+
+MetalLB is not deployed in every environment so it needs to be configurable
+from the Bootstrap configuration file, that’s why we have a new field about
```
Comment (NicolasT) [inline docs/developer/architecture/control-plane-ingress.rst:143]: How is this done? What's for configuration format? How is this backwards-compatible? How can it be made forward-compatible?

File: docs/developer/architecture/control-plane-ingress.rst [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+we will use a simple Layer2 configuration for MetalLB so that if the user
+environment allows using Layer2 protocol of MetalLB you just have to enable
+MetalLB, provide an Ingress IP for the Control Plane and that it.
+
+Default MetalLB configuration:
+
+.. code-block:: yaml
+
+    address-pools:
+    - name: ingress-ip
```
Comment (NicolasT) [inline docs/developer/architecture/control-plane-ingress.rst]: Let's find a more descriptive name.

File: docs/developer/architecture/control-plane-ingress.rst [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+      addresses:
+      - <control-plane ingress ip>/32
+      auto-assign: false
+
+But, since BGP router configuration can be really different between various
+environments, we need to make sure this
+`MetalLB configuration <https://metallb.universe.tf/configuration/#bgp-configuration>`_
+is exposed, that’s why we also have the ability to override default MetalLB
+configuration from the Bootstrap configuration file in order to provide BGP
+router information.
```
Comment (NicolasT) [inline docs/developer/architecture/control-plane-ingress.rst]: This implies we'd support BGP-based environments. Should we, now? Could we, later?

File: docs/developer/architecture/control-plane-ingress.rst [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+When we use MetalLB we do not want to use the same NGINX Ingress Controller
+deployments, since MetalLB will be the entry point in the Kubernetes cluster
+we do not need to use a DaemonSet running on every Control Plane nodes,
+instead, we will use a Deployment with 2 replicas.
+
+We also need to configure the Service for Ingress Controller differently
+depending on if we use MetalLB or not when we use it we want to use a
+LoadBalancer service and set the LoadBalancerIP to IngressIP provided by
+the user. If we do not use MetalLB then we use ClusterIP Service with IngressIP
+provided by the user as External IPs.
```
Comment (NicolasT) [inline docs/developer/architecture/control-plane-ingress.rst]: No mention of `externalTrafficPolicy`?

File: docs/developer/architecture/control-plane-ingress.rst [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+non-MetalLB we need to make sure the Salt states that deploy NGINX Ingress
+Controller remove no-longer-needed objects (e.g.: if you switch from
+non-MetalLB to MetalLB you want to remove the DaemonSet for NGINX Ingress
+Controller).
+
+Documentation
+-------------
+
+- Describe all new Bootstrap configuration fields (with some links
+  to MetalLB documentation to configure MetalLB to use BGP protocol)
```
Comment (NicolasT) [inline docs/developer/architecture/control-plane-ingress.rst]: So we'd directly expose (a subset of) the MetalLB configuration in `BootstrapConfiguration`? Not sure that's wise. What if we want to switch MetalLB to something else later? If it really required to have access to all of MetalLBs tunables? Is that a feature we provide?

File: docs/developer/architecture/control-plane-ingress.rst [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+- Add a simple procedure to change the Control Plane Ingress IP and
+  reconfigure all Control Plane conponents that need to.
+
+Test Plan
+---------
+
+Add some End-to-End tests in the CI:
+
+- Use MetalLB and a VIP as Control Plane Ingress IP
```
Comment (NicolasT) [inline docs/developer/architecture/control-plane-ingress.rst:201]: Likely needs a VIP failover test scenario as well.

**Review by NicolasT — DISMISSED** [review]

File: docs/developer/architecture/control-plane-ingress.rst (hunk 3 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: sure it's the plan :wink:

File: docs/developer/architecture/control-plane-ingress.rst (hunk 4 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst:143]: You can check #3418, basically, I have a `metallb` key in `networks:controlPlane` and in this dict I have `enabled` (which default to `false`)  and an optional `config` field that get merge with a default MetalLB configuration
https://github.com/scality/metalk8s/blob/a406960d634d9f6b06b0e0c770d3b673e1ae473f/salt/_pillar/metalk8s.py#L94-L119

File: docs/developer/architecture/control-plane-ingress.rst (hunk 5 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: Sure, any suggestion? 
Maybe `metalk8s-control-plane-entrypoint-ip` ?

File: docs/developer/architecture/control-plane-ingress.rst (hunk 6 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: We can say we will do this later, just it's really easy to expose this MetalLB functionally, but maybe we do not want to expose it for the moment and do not talk about it at all

File: docs/developer/architecture/control-plane-ingress.rst (hunk 7 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: Right, I will add it (btw it's done in implementation PR)

File: docs/developer/architecture/control-plane-ingress.rst (hunk 8 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: Depends if we want to support BGP or not, if we only support Layer2 mode then we do not need to expose MetalLB config and just add "something" to enable or not MetalLB (talking about MetaLLB directly in doc/bootstrap config or not)

File: docs/developer/architecture/control-plane-ingress.rst (hunk 9 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst:201]: Hmmm, right but not sure it's really easy to do, I don't know if we have an easy way to know where this VIP sit at any point in time (except doing an ARP request to get mac address and then checking the host having the network interface with this mac address)
Anyway I agree, I will add it here

File: docs/developer/architecture/control-plane-ingress.rst (hunk 4 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst:143]: But as I answered in another comment below, if we do not support BGP we do not need to expose config and we can also hide "metallb" from Bootstrap config and documentation, and just put `manageVIP: true/false` (or something else) in the bootstrap config

File: docs/developer/architecture/control-plane-ingress.rst (hunk 9 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst:201]: (Added)

File: docs/developer/architecture/control-plane-ingress.rst [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+Initially, Control Plane Ingress Controller was deployed using a DaemonSet with
+a ClusterIP service using Bootstrap Control Plane IP as External IP and then
+configuring all Control Plane components using this “External IP” (like OIDC
+and various UIs).
+
+So it means, we can only reach those components using Bootstrap Control
+Plane IP, and if, for whatever reason, Bootstrap is down you can no longer
+access any UIs and you need to reconfigure all components manually, change the
+IP used everywhere to another Control Plane node IP, in order to restore
+access to Control Plane components.
```
Comment (gdemonet) [inline docs/developer/architecture/control-plane-ingress.rst]: Or restore the Bootstrap :wink:

File: docs/developer/architecture/control-plane-ingress.rst (hunk 1 above)
Comment (gdemonet) [reply docs/developer/architecture/control-plane-ingress.rst]: But then we need to test it and make it configurable... If most of our customer base and target audience for ARTESCA is OK with the L2 approach, I think we should try to keep it constrained. Once we get a clear use-case to address, we can add it easily using your suggested approach :ok_hand:

File: docs/developer/architecture/control-plane-ingress.rst [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+--------------
+
+To have a proper HA Control Plane Ingress we want to use a
+`Virtual IP <https://en.wikipedia.org/wiki/Virtual_IP_address>`_ using
+`MetalLB <https://metallb.universe.tf/>`_ so that we can rely on layer2
+ARP requests when possible or let MetalLB advertise IPs using BGP when
+the user router support it.
+
+NOTE: In this document we talk about MetalLB used for Control Plane
+Ingress and not for Workload Plane.
```
Comment (gdemonet) [inline docs/developer/architecture/control-plane-ingress.rst]: Maybe remind why: we would likely need to go with "real" load-balancing, which makes the current implementation of MetalLB problematic (can only manage one VIP per Service, IIRC).

File: docs/developer/architecture/control-plane-ingress.rst [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+Instead of using a Virtual IP we could rely on DNS resolution use this
+FQDN to configure Control Plane components. In this case we let the user
+configure his DNS to resolve on an IP of one Control Plane node.
+
+WIth this approach we let the DNS server handle the “High Availability” of
+the Control Plane UIs, basically if we lose the node that resolve the DNS
+then we expect the DNS server to switch to another IP of a working Control
+Plane node.
+
+This approach was rejected because it’s not a real HA and we expect DNS server
```
Comment (gdemonet) [inline docs/developer/architecture/control-plane-ingress.rst]: > not a real HA

Meaning...? Something along the lines of "transparent for clients", which isn't the case for DNS-based failover?

File: docs/developer/architecture/control-plane-ingress.rst [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+you use MetalLB and will default to Bootstrap Control Plane IP if not
+(so that we have the same behavior as before).
+
+This Control Plane Ingress IP can be changed at any time just by editing
+the Bootstrap configuration file and follow a simple documented procedure
+with some Salt states to reconfigure every component that needs to be
+re-configured.
+
+NOTE: Changing this Control Plane Ingress IP means we need to reconfigure
+all Kubernetes APIServer since we use this Ingress IP as an OIDC provider.
```
Comment (gdemonet) [inline docs/developer/architecture/control-plane-ingress.rst:137]: this will change with oauth2-proxy I think, but we'll need to reconfigure the proxy anyway :innocent:

File: docs/developer/architecture/control-plane-ingress.rst (hunk 5 above)
Comment (gdemonet) [reply docs/developer/architecture/control-plane-ingress.rst]: `control-plane-ingress-ip`?

File: docs/developer/architecture/control-plane-ingress.rst [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+
+As for every other addon in MetalK8s, we will use the MetalLB helm chart and
+render this one using a specific “option” file. But this one will not be always
+deployed as we only want to deploy it when a specific key is set in the
+Bootstrap configuration file, so in the Salt pillar at the end.
+
+When we use MetalLB we do not want to use the same NGINX Ingress Controller
+deployments, since MetalLB will be the entry point in the Kubernetes cluster
+we do not need to use a DaemonSet running on every Control Plane nodes,
+instead, we will use a Deployment with 2 replicas.
```
Comment (gdemonet) [inline docs/developer/architecture/control-plane-ingress.rst:173]: Won't this cause unnecessary mesh networking? I guess it's not super critical for control plane, but I don't really see the benefit of switching to a Deployment.

File: docs/developer/architecture/control-plane-ingress.rst [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+
+We also need to configure the Service for Ingress Controller differently
+depending on if we use MetalLB or not when we use it we want to use a
+LoadBalancer service, set the LoadBalancerIP to IngressIP provided by
+the user and set externalTrafficPolicy to Local. If we do not use MetalLB
```
Comment (gdemonet) [inline docs/developer/architecture/control-plane-ingress.rst:178]: If we use `Local`, why don't we set up the Ingress controller pods to run locally to each node? What happens if no replica runs locally?

**Review by gdemonet — COMMENTED** [review]
##### TL;DR

Generally OK :ok_hand: with the design.
Simply questioning the deployment strategy change for the Ingress controller, with MetalLB in place (maybe I misunderstood something).
Also suggesting that we relegate BGP support to a later (optional) iteration.

File: docs/developer/architecture/control-plane-ingress.rst (hunk 10 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: True, but it's only in case of bootstrap node down for ever, I will add it

File: docs/developer/architecture/control-plane-ingress.rst (hunk 12 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: It may be transparent if the DNS server is active and properly switch from an IP to another in case of failure, but as is it's a "False HA" since we do not have any HA we just move the "problem": "we do not need the IP to be HA but a specific DNS"

File: docs/developer/architecture/control-plane-ingress.rst (hunk 13 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst:137]: Likely yes, but since we do not have any design for oauth2-proxy yet I do not mention it

File: docs/developer/architecture/control-plane-ingress.rst (hunk 1 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: Yes I agree, I can remove it for the moment and stay on Layer2 only. 
Maybe @xaf-scality you have an opinion on this one as well ?

File: docs/developer/architecture/control-plane-ingress.rst (hunk 11 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst]: Not sure it's worth talking about it here, it's really the design for "Control Plane Ingress"

File: docs/developer/architecture/control-plane-ingress.rst (hunk 14 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst:173]: It's less ressources, here we stick with 2 replicas for the full cluster (no matter the number of Control Plane nodes).
To me it's more like, "why should we use a DaemonSet when we can use a Deployment ?" 
I think @NicolasT also had other arguments

File: docs/developer/architecture/control-plane-ingress.rst (hunk 15 above)
Comment (TeddyAndrieux) [reply docs/developer/architecture/control-plane-ingress.rst:178]: Hmm to me, when you use Local anyway the VIP will sit on one node that has some "pod" of the Service running locally,so it cannot happens

File: docs/developer/architecture/control-plane-ingress.rst (hunk 11 above)
Comment (gdemonet) [reply docs/developer/architecture/control-plane-ingress.rst]: OK, then maybe promote this clear distinction to a "Goals vs. Non-goals" section? A note in the "Design Choices" seems too late to me.

File: docs/developer/architecture/control-plane-ingress.rst (hunk 15 above)
Comment (gdemonet) [reply docs/developer/architecture/control-plane-ingress.rst:178]: Ah OK, I misunderstood this behaviour then.

File: docs/developer/architecture/control-plane-ingress.rst (hunk 14 above)
Comment (gdemonet) [reply docs/developer/architecture/control-plane-ingress.rst:173]: OK OK, I would just vote for lowering the priority of this switch:

- it's an architecture change, so it has an impact on users willing to operate/monitor our stuff
- it's not a huge win (an idle Ingress controller doesn't take much resources)

File: docs/developer/architecture/control-plane-ingress.rst [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+This approach was rejected because Control Plane components we deploy today
+in MetalK8s does not seems to all support relative path and it does not
+seems to be possible to have proper HA without re-configuring some Control
+Plane components when we lose a Control Plane node.
+
+Non-Goals
+---------
+
+Those points may be addresses later (or not) but in this document, we focus on
```
Comment (gdemonet) [inline docs/developer/architecture/control-plane-ingress.rst]:
```suggestion
These points may be addressed later (or not) but in this document, we focus on
```

File: docs/developer/architecture/control-plane-ingress.rst [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+with some Salt states to reconfigure every component that needs to be
+re-configured.
+
+NOTE: Changing this Control Plane Ingress IP means we need to reconfigure
+all Kubernetes APIServer since we use this Ingress IP as an OIDC provider.
+
+MetalLB Configuration
+~~~~~~~~~~~~~~~~~~~~~
+
+MetalLB is not deployed in every environment so it needs to be activatable
```
Comment (gdemonet) [inline docs/developer/architecture/control-plane-ingress.rst]:
```suggestion
MetalLB is not deployed in every environment so it needs to be enabled
```

**Review by gdemonet — DISMISSED** [review]
(minor typos) :shipit:

**Review by gdemonet — APPROVED** [review]
